### PR TITLE
feat(retry): support a exponential backoff retry mechanism

### PR DESCRIPTION
**Pull request by daddykotex**

### CODE REVIEW

File: spec/backoff/backoff.spec.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      }
+    }
+  });
+  server.listen(9615);
+
+  afterAll(function() {
+    server.close();
+  });
+
+  it('retry and stop and a successful response is returned', function(done) {
```
Comment (jtheoof) [inline spec/backoff/backoff.spec.js]: missing `should` in all `it` statements (for consistency with other tests)

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,863 @@
+{
```
Comment (jtheoof) [inline package-lock.json]: why was this file introduced?

File: src/http/backends/node_backend.ts [hunk 3]
```diff
@@ -55,10 +55,14 @@ export function nodeHttpRequest(request: Request): Promise<any> {
...
 
-        if (response.statusCode >= 200 && response.statusCode < 300) {
+        const statusCode = response.statusCode;
+        if (statusCode >= 200 && statusCode < 300) {
           resolve(payload);
         } else {
-          reject(payload);
+          reject({
+            statusCode,
+            payload
```
Comment (jtheoof) [inline src/http/backends/node_backend.ts]: How can this be a valid JS object? 
```
{
  503,
  {...},
}
```
Is not a valid object. Not sure how come typescript is not complaining.

File: src/mnubo.ts [hunk 4]
```diff
@@ -168,19 +176,45 @@ export class Client {
...
+        retries: exponentialBackoff.numberOfAttempts,
+        minTimeout: exponentialBackoff.initialDelayInMillis,
+        randomize: true
+      };
+
+      return promiseRetry((retry: any, attempt: number) => {
+        return f().catch((err) => {
+          if (err.statusCode === 503) {
+            if (attempt <= exponentialBackoff.numberOfAttempts) {
+              exponentialBackoff.onRetry(attempt);
```
Comment (jtheoof) [inline src/mnubo.ts]: What if `onRetry` has not been passed in the options?

File: src/mnubo.ts [hunk 5]
```diff
@@ -168,19 +176,45 @@ export class Client {
...
+      const opts: any = {
+        retries: exponentialBackoff.numberOfAttempts,
+        minTimeout: exponentialBackoff.initialDelayInMillis,
+        randomize: true
+      };
+
+      return promiseRetry((retry: any, attempt: number) => {
+        return f().catch((err) => {
+          if (err.statusCode === 503) {
+            if (attempt <= exponentialBackoff.numberOfAttempts) {
```
Comment (jtheoof) [inline src/mnubo.ts]: What happens when the threshold is reached? I don't see the `else`.

File: spec/backoff/backoff.spec.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+const mnubo = require('../../dist/mnubo');
+const http = require('http');
+const port = 9615;
+
+describe('backoff', function() {
+  jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
+
+  let serverCounter = {};
+  let serverLimit = {};
+  const server = http.createServer((req, res) => {
```
Comment (jtheoof) [inline spec/backoff/backoff.spec.js]: Very clever.

I would improve readability by wrapping all of this inside its own class.

```js
class BackoffSimulatorServer {
    constructor(limit) {
      this.counter = {};
      this.limit = limit || {};
      this.initServer();
    }

    addLimit(url) {
      this.limit[url] = 0;
    }

    initServer() {
      this.server = ...// Your server code.
    }

    ...
}
```

What do you think?

File: src/http/backends/node_backend.ts (hunk 3 above)
Comment (blemoine) [inline src/http/backends/node_backend.ts]: hmmm yes it it a valid object, it's equivalent to:
{ statusCode:503, etc.} for example

File: package-lock.json (hunk 2 above)
Comment (daddykotex) [inline package-lock.json]: `npm i` logged :
successfully installed, generated package-lock.json to be committed

`yarn` does this

File: src/http/backends/node_backend.ts (hunk 3 above)
Comment (daddykotex) [inline src/http/backends/node_backend.ts]: it gives:
```
{
   "statusCode": 503,
   "payload": payload
}
```

File: src/mnubo.ts (hunk 5 above)
Comment (daddykotex) [inline src/mnubo.ts]: the library is weirdly implemented, but it the logic that determine if a retry should be attempted is happening in the `retry()` function below.

File: src/mnubo.ts (hunk 4 above)
Comment (daddykotex) [inline src/mnubo.ts]: good question, i did not do check

File: spec/backoff/backoff.spec.js (hunk 1 above)
Comment (daddykotex) [inline spec/backoff/backoff.spec.js]: will do

File: spec/backoff/backoff.spec.js (hunk 6 above)
Comment (daddykotex) [inline spec/backoff/backoff.spec.js]: 👍

File: src/http/backends/node_backend.ts (hunk 3 above)
Comment (jtheoof) [inline src/http/backends/node_backend.ts]: Forgive my ignorance.

Comment (jtheoof) [inline src/http/backends/node_backend.ts]: http://es6-features.org/#PropertyShorthand

File: spec/backoff/backoff.spec.js (hunk 6 above)
Comment (daddykotex) [inline spec/backoff/backoff.spec.js]: done

File: spec/backoff/backoff.spec.js (hunk 1 above)
Comment (daddykotex) [inline spec/backoff/backoff.spec.js]: done

File: README.md [hunk 7]
```diff
@@ -34,7 +34,8 @@ The initialization of the SDK client requires two mandatory fields:
 - `id` (**mandatory**): The client id.
 - `secret` (**mandatory**): The client secret.
 - `env` (*optional*): The environment where the API calls will be sent. It can be either `sandbox` or `production`. By default the `env` is `sandbox`.
-- `compression` (*optional*): If you want to use `gzip` compression for both inbound and outbound data, set the value to `true`. By default this parameter is `false.
+- `compression` (*optional*): If you want to use `gzip` compression for both inbound and outbound data, set the value to `true`. By default this parameter is `false`.
+- `exponentialBackoff` (*optional*): If you want to, automatically, retry (with a delay) when you receive 503 from mnubo. By default, this parameter is `false`.
```
Comment (jtheoof) [inline README.md]: It looks like the parameters to pass are a bit more complex than `true` or `false`.

```
exponentialBackoff: {
    numberOfAttempts: retries,
    initialDelayInMillis: delay,
    onRetry: onRetryFn
}
```

Or am I missing something?

**Comment by jtheoof** [comment]
Looks almost good, one question regarding the doc.

**Comment by daddykotex** [comment]
In fact, I wrote the doc before implementing the feature, and realized that it would be easier to avoid `boolean` here

File: README.md (hunk 7 above)
Comment (daddykotex) [inline README.md]: done

**Pull request merged**